### PR TITLE
Support oblique perspective projection in ClipPlanesEvaluator

### PR DESCRIPTION
**Pull request by atomicsulfate**

### CODE REVIEW

File: @here/harp-mapview/lib/CameraUtils.ts [hunk 1]
```diff
@@ -4,33 +4,128 @@
...
+        return halfSide / Math.tan(fov / 2);
+    }
+
+    const ppOffsetSq = ppOffset ** 2;
+
+    // Special case for (close to) right angle fov, tangent approaches infinity.
+    const eps = 1e-6;
+    if (Math.abs(fov - Math.PI / 2) < eps) {
+        // docs
+        return halfSide * Math.sqrt(1 - ppOffsetSq);
```
Comment (nzjony) [inline @here/harp-mapview/lib/CameraUtils.ts:76]: If the angle is close to a right angle, won't the focal length be (almost) 0?

File: @here/harp-mapview/lib/CameraUtils.ts [hunk 2]
```diff
@@ -40,8 +135,20 @@ export namespace CameraUtils {
      * @param camera
      * @param hFov - The horizontal field of view in radians.
      */
-    export function setHorizontalFov(camera: THREE.PerspectiveCamera, hFov: number): void {
-        camera.fov = THREE.MathUtils.radToDeg(2 * Math.atan(Math.tan(hFov / 2) / camera.aspect));
+    export function setHorizontalFov(
+        camera: THREE.PerspectiveCamera,
+        hFov: number,
+        width: number
```
Comment (nzjony) [inline @here/harp-mapview/lib/CameraUtils.ts:141]: missing param

File: @here/harp-mapview/lib/CameraUtils.ts [hunk 3]
```diff
@@ -4,33 +4,128 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+import { Vector2Like } from "@here/harp-geoutils";
+import { assert } from "@here/harp-utils";
 import * as THREE from "three";
 
 import { MAX_FOV_RAD, MIN_FOV_RAD } from "./FovCalculation";
 
+function isCenteredProjection(ppalPoint: Vector2Like): boolean {
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/CameraUtils.ts]: changed to principalPoint.

File: @here/harp-mapview/lib/CameraUtils.ts [hunk 4]
```diff
@@ -4,33 +4,128 @@
...
+    return ppalPoint.x === 0 && ppalPoint.y === 0;
+}
+
+function computePosSideFov(focalLength: number, ppOffset: number, viewportSide: number): number {
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/CameraUtils.ts:26]: documentation added .

File: @here/harp-mapview/lib/CameraUtils.ts [hunk 5]
```diff
@@ -4,33 +4,128 @@
...
+    return Math.atan(((1 - ppOffset) * viewportSide * 0.5) / focalLength);
+}
+
+function computeNegSideFov(focalLength: number, ppOffset: number, viewportSide: number): number {
+    return Math.atan(((1 + ppOffset) * viewportSide * 0.5) / focalLength);
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/CameraUtils.ts]: removed, as you say it's enough with the positive side and changing the offset sign.

File: @here/harp-mapview/lib/CameraUtils.ts [hunk 6]
```diff
@@ -4,33 +4,128 @@
...
+    ppalPoint: Vector2Like,
+    focalLength: number,
+    height: number,
+    hFov: number
+): void {
+    if (isCenteredProjection(ppalPoint)) {
+        delete camera.userData.fovs;
+    } else {
+        const width = height * camera.aspect;
+        camera.userData.fovs = {
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/CameraUtils.ts:77]: added

File: @here/harp-mapview/lib/CameraUtils.ts [hunk 7]
```diff
@@ -40,8 +135,20 @@ export namespace CameraUtils {
...
+        width: number
+    ): void {
+        const focalLength = computeFocalLengthFromHorizontalFov(camera, hFov, width);
+        camera.fov = isCenteredProjection(getPrincipalPoint(camera))
+            ? THREE.MathUtils.radToDeg(2 * Math.atan(Math.tan(hFov / 2) / camera.aspect))
+            : computeVerticalFov(camera, focalLength, width / camera.aspect);
+        setFocalLength(camera, focalLength);
+    }
+
+    export function getFocalLength(camera: THREE.PerspectiveCamera): number {
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/CameraUtils.ts]: done

File: @here/harp-mapview/lib/CameraUtils.ts [hunk 8]
```diff
@@ -4,33 +4,128 @@
...
+            horizontal: hFov
+        };
+    }
+}
+
+function setFocalLength(camera: THREE.PerspectiveCamera, focalLength: number): void {
+    camera.userData.focalLength = focalLength;
+}
+
+function vertToHorizSymFov(camera: THREE.PerspectiveCamera): number {
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/CameraUtils.ts]: removed

File: @here/harp-mapview/lib/CameraUtils.ts [hunk 9]
```diff
@@ -4,33 +4,128 @@
...
+}
+
+/**
+ * Computes a camera's focal length from vertical fov and viewport height or horizontal fov and
+ * viewport width.
+ * @beta
+ *
+ * @param fov - Vertical or horizontal field of view in radians.
+ * @param viewportSide - Viewport height if fov is vertical, otherwise viewport width.
+ * @param ppOffset - Principal point offset in y direction if fov is vertical,
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/CameraUtils.ts:93]: added

File: @here/harp-mapview/lib/CameraUtils.ts (hunk 1 above)
Comment (atomicsulfate) [reply @here/harp-mapview/lib/CameraUtils.ts:140]: I've added quite some documentation explaining everything, hope it's clearer now.

File: @here/harp-mapview/lib/CameraUtils.ts (hunk 2 above)
Comment (atomicsulfate) [reply @here/harp-mapview/lib/CameraUtils.ts]: function removed.

**Comment by atomicsulfate** [comment]
> You should mention that this is in NDC space and not pixels. 

done

File: @here/harp-mapview/lib/CameraUtils.ts [hunk 10]
```diff
@@ -4,74 +4,244 @@
...
+// In centered projections the principal point is at NDC origin, splitting vertical and horizontal
+// fovs in two equal halves.
+function isCenteredProjection(principalPoint: Vector2Like): boolean {
+    return principalPoint.x === 0 && principalPoint.y === 0;
+}
+
+/**
+ * Computes the fov on the positive side of NDC x or y dimension (i.e. either right or top fov).
+ * @param focalLength - Focal length in pixels. It must be larger than 0.
+ * @param ppOffset - Principal point offset either in y or x dimension.
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/CameraUtils.ts:22]: done

File: @here/harp-mapview/lib/CameraUtils.ts [hunk 11]
```diff
@@ -4,74 +4,244 @@
...
+function computePosSideFov(focalLength: number, ppOffset: number, viewportSide: number): number {
+    // see diagram in computeFocalLengthFromFov().
+    assert(focalLength > 0, "Focal length must be larger than 0");
+    return Math.atan(((1 - ppOffset) * viewportSide * 0.5) / focalLength);
+}
+
+/**
+ * Computes the vertical or horizontal fov.
+ * @param focalLength - Focal length in pixels. It must be larger than 0.
+ * @param ppOffset - Principal point offset in y (vertical fov) or x dimension (horizontal fov).
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/CameraUtils.ts:35]: done

File: @here/harp-mapview/lib/CameraUtils.ts [hunk 12]
```diff
@@ -4,74 +4,244 @@
...
+              computePosSideFov(focalLength, -ppOffset, viewportSide);
+}
+
+/**
+ * For off-center projections, fovs are asymmetric. In that case, the camera saves some of the side
+ * fovs to save some computations. All values are in radians.
+ */
+interface Fovs {
+    top: number;
+    right: number;
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/CameraUtils.ts:55]: horizontal = left + right, added as a comment. Saving those 3 fovs we can get the rest by subtraction, see CameraUtils.getLeftFov() and CameraUtils.getBottomFov().

File: @here/harp-mapview/lib/CameraUtils.ts [hunk 13]
```diff
@@ -4,74 +4,244 @@
...
+ * @beta
+ *
+ * @param fov - Vertical or horizontal field of view in radians.
+ * @param viewportSide - Viewport height if fov is vertical, otherwise viewport width.
+ * @param ppOffset - Principal point offset in y direction if fov is vertical,
+ * otherwise in x direction.
+ * @returns focal length in pixels.
+ */
+function computeFocalLengthFromFov(fov: number, viewportSide: number, ppOffset: number): number {
+    //               C <- Projection center
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/CameraUtils.ts:98]: thanks!

File: @here/harp-mapview/lib/CameraUtils.ts [hunk 14]
```diff
@@ -4,74 +4,244 @@
...
+        // a^2 = f^2 + (1-ppOff)*s/2
+        // b^2 = f^2 + (1+ppOff)*s/2
+        // h^2 = a^2 + b^2
+        // Substitute a^2 and b^2 in third equation and solve for f to get:
+        // f = (s/2) * sqrt(1-ppOff^2)
+        return halfSide * Math.sqrt(1 - ppOffsetSq);
+    }
+
+    // Case 2b. General asymmetric fov case:
+    // (1)   tan(pfov) = (1-ppOff)*s / (2*f)
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/CameraUtils.ts:144]: You only know pfov = nfov = fov /2 for centered projections. For an off-center projection, you don't know how the fov is split into pfov and nfov. So you have 3 unknowns pfov, nfov and f, and therefore need the three equations.

File: @here/harp-mapview/lib/CameraUtils.ts (hunk 1 above)
Comment (nzjony) [reply @here/harp-mapview/lib/CameraUtils.ts:140]: Thanks, I was confused, now it makes sense.

File: @here/harp-mapview/lib/CameraUtils.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (nzjony) [reply @here/harp-mapview/lib/CameraUtils.ts:144]: Thanks for the explanation, I missed that you need to compute two things at once, thanks.

**Review by nzjony — APPROVED** [review]

**Pull request merged**